LSPS5/README.md

```diff
@@ -0,0 +1,623 @@
+the JSON string `"2.0"`.
+
+`method` is the required name of the notification, and MUST
+be one of the webhook notification methods listed later in
+this specification.
+The `method` specifies exactly what the LSP wants to notify
+to the client as a push notification.
+
+`params` is a required by-name parameter object for the
+notification method.
+The parameters may be empty, if the method allows the
+parameters to be empty.
+
+The LSP creates a timestamp of when the notification is
+created, then signs both the timestamp and the above body.
+
+The signature is by the LSP using its node ID for signing
+as described in [<LSPS0 ln_signature>][].
+
+[<LSPS0 ln_signature>]: ../LSPS0/common-schemas.md#link-lsps0ln_signature
+
+The message to be signed is the following JSON string template:
+
+    "LSPS5: DO NOT SIGN THIS MESSAGE MANUALLY: LSP: At ${timestamp} I notify ${body}"
+
+Where `${timestamp}` is the timestamp, in ISO8601
+date format `YYYY-MM-DDThh:mm:ss.uuuZ` [<LSPS0 datetime>][],
+and `${body}` is the exact and complete JSON serialization of
+the above webhook notification object.
+
+[<LSPS0 datetime>]: ../LSPS0/common-schemas.md#link-lsps0datetime
+
+For example, if the timestamp is `2023-05-04T10:52:58.395Z`
+and the webhhok notification object is the exact ASCII sequence
+`{"jsonrpc":"2.0","method":"lsps5.goodbye","params":{}}`,
+then the JSON string to be signed is:
+
+    "LSPS5: DO NOT SIGN THIS MESSAGE MANUALLY: LSP: At 2023-05-04T10:52:58.395Z I notify {\"jsonrpc\":\"2.0\",\"method\":\"lsps5.goodbye\",\"params\":{}}"
+
+The message to be signed is the contained string in UTF-8
+format, without a trailing `NUL` character.
+For the above example, the message to be signed would have the
+following hex dump:
+
+```
+00000000: 4c53 5053 353a 2044 4f20 4e4f 5420 5349  LSPS5: DO NOT SI
+00000010: 474e 2054 4849 5320 4d45 5353 4147 4520  GN THIS MESSAGE
+00000020: 4d41 4e55 414c 4c59 3a20 4c53 503a 2041  MANUALLY: LSP: A
+00000030: 7420 3230 3233 2d30 352d 3034 5431 303a  t 2023-05-04T10:
+00000040: 3532 3a35 382e 3339 355a 2049 206e 6f74  52:58.395Z I not
+00000050: 6966 7920 7b22 6a73 6f6e 7270 6322 3a22  ify {"jsonrpc":"
+00000060: 322e 3022 2c22 6d65 7468 6f64 223a 226c  2.0","method":"l
+00000070: 7370 7335 2e67 6f6f 6462 7965 222c 2270  sps5.goodbye","p
+00000080: 6172 616d 7322 3a7b 7d7d                 arams":{}}
+```
+
+The resulting signature is a zbase32-formatted string.
+
+The timestamp, signature, and webhook notification object are sent
+to the notification delivery service based on the protocol of the
+webhook:
+
+* For HTTPS webhooks, the LSP makes a `POST` request, with
+  the webhook notification object as the exact `POST` request
+  body, and with additional HTTP headers below:
+  * `x-lsps5-timestamp` - the LSP-side timestamp, in ISO8601
+     date format `YYYY-MM-DDThh:mm:ss.uuuZ`, for example
+    `2023-05-04T10:14:23.853Z`.
+  * `x-lsps5-signature` - the LSP-generated signature, as a
+    zbase32 string.
```

Comment:
Can we add the `lsp node id` + `client node id` as a http header and also add `client node id` to the signature. This information should always be required anyway + it makes the webhook url smaller.

Reply:
Right right right, so now ***ANY*** J. Random Hacker who reads your open-source client code can spam you by making up an LSP node ID and finding out your client node ID (which is probably cleartext on some invoice you are giving out) can spam you and the notification delivery service will innocently give the notifications to the client.

We need some kind of "token" by which the client informs the notification delivery server "this LSP node ID is good for this client node ID".  And ANY passthrough from client->LSP->notification delivery server can be done via the URL.  Just put the relevant information in the URL.

Bonus points is that the client can have a separate API in the notification delivery server that registers the client ID and LSP ID and gets back a smaller token that is put in the URL.

Reply:
So lsp node id + client node id + authorization (aka signature) in http header? These are things ever NDS needs so why not standardize it? 

What is your motivation to put everything in the url and therefore leave the design space to the app developers? This just leaves room for error IMO.

Reply:
Less things that need to be specified, mostly.

For example, the client and notification delivery server could have their own communications channel by which the client presents its authorization + LSP node ID + client node ID to the notification delivery service directly, then gets back a smaller token that it just adds as `?token=whateverness` to the end of the URL, rather than always passing this information to the LSP.

Reply:
@SeverinAlexB I don't think the NDS needs to know the client node id?

Reply:
> Can you elaborate on that? Do you want to use a device ID or so? How would multi device work then in this case?

Yeah basically use a device id. That's what you need in order to send a notification to the user's app. That's the job of the notification delivery service, to send these notifications to the user's device. This data, how to reach the client, is encoded in the client's webhook url (somehow). The node id then is not necessary to know. The device id is enough.

Reply:
So in case Bitkit is running on my mobile phone and desktop at the same time, it would register 2 webhooks?

Reply:
> So in case Bitkit is running on my mobile phone and desktop at the same time, it would register 2 webhooks?

You could do with one webhook if you want, as long as Bitkit's NDS is aware which devices to notify based on the url. You could let the devices register themselves to the NDS, and notify both devices when the webhook comes in for example.

Either case, the node id is not necessary?

Reply:
> I see, agree. Do you have the implicit assumption to make the node id optional because of privacy and tho prevent the NDS as a honeypot or so?

I do, yes. If there is no reason to include the node id I'd prefer the NDS to be unaware of it wherever possible.

> If yes, I missed this and maybe it's worth a reasoning section in the spec?

I don't think I ever mentioned this before. Adding the reasoning to the spec doesn't sound unreasonable.